### DISCUSSION TITLE
[Improvement]: Unify configuration environment and class definition variables

### DISCUSSION THREAD

**Discussion by mysliwietzflorian** [pimcore/pimcore#12655] [General]
While I love the new possibilities with the Pimcore [configuration environments](https://pimcore.com/docs/pimcore/10.5/Development_Documentation/Deployment/Configuration_Environments.html) and the [class definition](https://pimcore.com/docs/pimcore/10.5/Development_Documentation/Deployment/Deployment_Tools.html#page_Pimcore-Class-Definitions) configs I think there are some inconsistencies that I wanted to address. I refer to the current docs (dev).

- **[default writability]** For the configs the default directory is `/var/config/` (same as with class definitions). Moving them into `config/pimcore/...` will not have any effect on writability. The class definitions will be read-only when moved into `config/pimcore/classes`. IMO it should be the same logic as for the configs.

- **[directory location]** The environment variable to change the storage variable for configs is `PIMCORE_CONFIG_STORAGE_DIR_<type>` (e.g.: `PIMCORE_CONFIG_STORAGE_DIR_IMAGE_THUMBNAILS`). For class definitions it is `PIMCORE_CLASS_DEFINITION_DIRECTORY `. Can we unify this?

- **[read-only flag]** To make configs read-only you set `PIMCORE_WRITE_TARGET_<type>=disabled`. For class definitions it is `PIMCORE_CLASS_DEFINITION_WRITABLE=0`. Can we unify this? 

- **[write target]** For configs we have `symfony-config` (yaml) and `settings-store` (database) on hand. For the class definitions only PHP configuration files. Is there a deeper reason why this wasn't deprecated and also moved to the other formats? Is this change planned?

- **[docs referencing each other]** I almost missed the possibilities for class definition configuration. Should we reference this on the [configuration environments](https://pimcore.com/docs/pimcore/10.5/Development_Documentation/Deployment/Configuration_Environments.html) page? Or do you feel they are two distinct aspects within Pimcore?

- **[docs config creation location ]** I could not find any information on where new configurations or class definitions are generated/created when you set a storage directory. Should we add this to the docs?

Can I help with some of these changes? I am glad for any information or hints about one of the points listed above. 😉 

Also see:
- https://github.com/pimcore/pimcore/pull/10812
- https://github.com/pimcore/pimcore/pull/12392

**Comment by kingjia90** [comment]
Thank you for the detailed recap and the good inputs!

We actually had similar thoughts and attempted to move in that direction but faced in some "problems".
In short and iirc, the class definition couldn't be easily unified by moving to LocationAwareConfigRepository (which allows to use db and yml, using same env as the others) without BC breaking or major refactoring due the nature of the class definition being `include` and initialized by using  `_set_state`. Also faced some issues with Custom Layouts about `setCondition`.

So for the moment, these were just "postponed" to Pimcore 11, but the good news is that after release of the 10.5 next week, the work in progress for 11 will soon begin. In case, let's see if it will be possible to assign you some of these related tasks 😄 

If you have any suggestions to improve documentation in the meanwhile, feel free to open PRs.

**Reply by mysliwietzflorian** [reply]
@kingjia90 Thanks for your response! Nice to hear!

Just to collect further points to keep in mind with the configurations:  
- **[DataHub]** uses the same configuration storage (write target) as the other configs. In the production environment, the default write-target is *disabled* (read-only; same as with other configs). The default directory location for the Symfony config (YAML) is `/var/config/data-hub/<name>.yaml` and cannot be changed (unlike the other configs). 

See https://pimcore.com/docs/data-hub/current/Deployment.html
